pkg/kubelet/util/pluginwatcher/example_plugin.go

```diff
@@ -0,0 +1,147 @@
+	server *examplePlugin
+}
+
+func (s *pluginServiceV1Beta1) GetExampleInfo(ctx context.Context, rqt *v1beta1.ExampleRequest) (*v1beta1.ExampleResponse, error) {
+	glog.Infof("GetExampleInfo v1beta1field: %s", rqt.V1Beta1Field)
+	return &v1beta1.ExampleResponse{}, nil
+}
+
+func (s *pluginServiceV1Beta1) RegisterService() {
+	v1beta1.RegisterExampleServer(s.server.grpcServer, s)
+}
+
+type pluginServiceV1Beta2 struct {
+	server *examplePlugin
+}
+
+func (s *pluginServiceV1Beta2) GetExampleInfo(ctx context.Context, rqt *v1beta2.ExampleRequest) (*v1beta2.ExampleResponse, error) {
+	glog.Infof("GetExampleInfo v1beta2_field: %s", rqt.V1Beta2Field)
+	return &v1beta2.ExampleResponse{}, nil
+}
+
+func (s *pluginServiceV1Beta2) RegisterService() {
+	v1beta2.RegisterExampleServer(s.server.grpcServer, s)
+}
+
+// NewExamplePlugin returns an initialized examplePlugin instance
+func NewExamplePlugin() *examplePlugin {
+	return &examplePlugin{
+		stop: make(chan bool),
+	}
+}
+
+// NewTestExamplePlugin returns an initialized examplePlugin instance for testing
+func NewTestExamplePlugin() *examplePlugin {
+	return &examplePlugin{
+		stop:               make(chan bool),
+		registrationStatus: make(chan registerapi.RegistrationStatus),
+	}
+}
+
+// GetInfo is the RPC invoked by plugin watcher
+func (e *examplePlugin) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
+	return &registerapi.PluginInfo{
+		Type:              PluginType,
+		Name:              PluginName,
+		SupportedVersions: []string{"v1beta1", "v1beta2"},
+	}, nil
+}
+
+func (e *examplePlugin) NotifyRegistrationStatus(ctx context.Context, status *registerapi.RegistrationStatus) (*registerapi.RegistrationStatusResponse, error) {
+	glog.Infof("NotifyRegistrationStatus")
+	if e.registrationStatus != nil {
+		e.registrationStatus <- *status
+	}
+	if !status.PluginRegistered {
+		glog.Errorf("Registration failed: %s\n", status.Error)
+	}
+	return &registerapi.RegistrationStatusResponse{}, nil
+}
+
+// Serve starts example plugin grpc server
+func (e *examplePlugin) Serve(socketPath string) error {
+	glog.Infof("starting example server at: %s\n", socketPath)
+	lis, err := net.Listen("unix", socketPath)
+	if err != nil {
+		glog.Errorf("starting example server failed: %v", err)
+		return err
+	}
+	glog.Infof("example server started at: %s\n", socketPath)
+	e.grpcServer = grpc.NewServer()
+	// Registers kubelet plugin watcher api.
+	registerapi.RegisterRegistrationServer(e.grpcServer, e)
+	// Registers services for both v1beta1 and v1beta2 versions.
+	v1beta1 := &pluginServiceV1Beta1{server: e}
+	v1beta1.RegisterService()
+	v1beta2 := &pluginServiceV1Beta2{server: e}
+	v1beta2.RegisterService()
+
+	// Starts service
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// Blocking call to accept incoming connections.
+		err := e.grpcServer.Serve(lis)
+		glog.Errorf("example server stopped serving: %v", err)
+	}()
+
+	wg.Wait()
```

Comment:
wg.Wait should probably in the `Stop()` call and you probably should timeout on the Wait that would remove the stopChan

Reply:
umm.. thats also possible but in that case will have to add `wg` to the watcher. Dont see much difference overall.

Reply:
Sure you can actually check for an error without starting a goroutine.
Usage of this function doesn't require you to start a waitgroup in the tests.

Code changes from:
```golang
var wg sync.WaitGroup
wg.Add(1)
go func() {
	require.NotNil(t, p.Serve(socketPath))
	wg.Done()
}()
c := make(chan struct{})
go func() {
	defer close(c)
	wg.Wait()
}()
select {
case <-c:
case <-time.After(time.Second):
	t.Fatalf("Timed out while on Serve()")
}
```

To:
```golang
require.NotNil(t, p.Serve(socketPath))
require.NotNil(t, p.Stop())
```